monero-serai: fee calculation parity with Monero's wallet2

j-berman opened the pull request:
The goal of this PR is to ensure `monero-serai` constructs transactions using fees that are indistinguishable from Monero's wallet2. This PR accomplishes this goal to the best of my knowledge. I haven't tested it with mainnet yet but plan on doing so (while working on the decoy selection algorithm next).

The core internal difference this PR introduces to `monero-serai` transaction construction logic is that decoys must be selected *before* first calculating the fee. This is because a ring's `key_offsets` can vary in byte length between decoy selection attempts. By fixing the selected decoys at the start of tx construction, the tx builder can immediately calculate the tx's weight, calculate the fee based on that weight, and settle on the user's change.

Of note, there is a theoretical discrepancy to wallet2 that I noted in a comment in the function `calculate_weight_and_fee`. This discrepancy should never apply in practice. In short, wallet2 can theoretically use an estimate that is significantly different from what the calculated tx fee should be, which could cause wallet2 to use a fee that is not calculated from the final tx's weight. This does not occur in practice to the best of my knowledge; wallet2 *always* under-estimates the fee today by a small margin (it could have happened in the past when there were very few outputs in the chain). The reason I kept this theoretical discrepancy in `monero-serai` is because handling it would require copying brittle code from wallet2 that has oddities (and what look like bugs) that would be a pain to integrate cleanly for no practical gain (e.g. wallet2 does not factor in the expected final weight of tx extra in its initial estimate, while the CLI, RPC, and GUI do factor in the estimated weight of a payment ID in tx extra only if sending to an integrated address).

j-berman committed:
monero-serai: fee calculation parity with Monero's wallet2

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +513,27 @@ impl<R: RpcConnection> Rpc<R> {
+    let priority_idx =
+      if priority >= 4 { 3 } else { priority.saturating_sub(1).try_into().unwrap() };
+
+    debug_assert!(priority_idx < res.fees.len(), "Unexpected size of fee response");
```
This allows a malicious node to crash any app with debug assertions. Please use an error here since this can't be stated as an invariant.

j-berman replied:
Fair.

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +513,27 @@ impl<R: RpcConnection> Rpc<R> {
+
+    debug_assert!(priority_idx < res.fees.len(), "Unexpected size of fee response");
+
+    // Using core Monero wallet2.cpp defaults if it can't use a fee from response
```
Mind explaining the safety of this? Should we just return an error instead?

j-berman replied:
I was trying to match wallet2's behavior where if the RPC response is bad, it returns these defaults. I'm fine with erroring here instead. Easier to reason through and it's a sensible error case.

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +513,27 @@ impl<R: RpcConnection> Rpc<R> {
+
+    // Using core Monero wallet2.cpp defaults if it can't use a fee from response
+    if priority_idx >= res.fees.len() {
+      Ok(Fee { per_weight: 300000 /* wallet2 FEE_PER_BYTE */, mask: 1 })
```
`res.fees.get(priority_idx).unwrap_or(300000)`

Also, is this meant to be static for all priorities? And a mask of 1???

j-berman replied:
I'm fine with erroring here instead as noted above

kayabaNerve commented on coins/monero/src/transaction.rs:
```diff
@@ -164,12 +165,13 @@ pub struct TransactionPrefix {
 
 impl TransactionPrefix {
-  pub(crate) fn fee_weight(ring_len: usize, inputs: usize, outputs: usize, extra: usize) -> usize {
+  pub(crate) fn fee_weight(decoy_weights: &Vec<usize>, outputs: usize, extra: usize) -> usize {
```
&[], not &Vec

kayabaNerve commented on coins/monero/src/transaction.rs:
```diff
@@ -214,12 +216,18 @@ pub struct Transaction {
 impl Transaction {
   pub(crate) fn fee_weight(
     protocol: Protocol,
+    decoy_weights: &Vec<usize>,
```
&[], not &Vec<>

kayabaNerve commented on coins/monero/src/transaction.rs:
```diff
@@ -214,12 +216,18 @@ pub struct Transaction {
   ) -> usize {
-    TransactionPrefix::fee_weight(protocol.ring_len(), inputs, outputs, extra) +
-      RctSignatures::fee_weight(protocol, inputs, outputs)
+    if inputs != decoy_weights.len() {
```
Replace inputs with decoy_weights.len()? Why take in the variable twice if it just introduces a potential inconsistency?

j-berman replied:
> Why take in the variable twice if it just introduces a potential inconsistency?

Because I didn't think of this here, good spot :)

kayabaNerve commented on coins/monero/src/wallet/decoys.rs:
```diff
@@ -16,6 +16,8 @@ use crate::{
   rpc::{RpcError, RpcConnection, Rpc},
 };
 
+use crate::serialize::{varint_len};
```
Please merge this use with the prior use crate.

kayabaNerve commented on coins/monero/src/wallet/decoys.rs:
```diff
@@ -131,7 +133,12 @@ pub struct Decoys {
 
+#[allow(clippy::len_without_is_empty)]
 impl Decoys {
+  pub fn fee_weight(offsets: &Vec<u64>) -> usize {
```
&[], not &Vec<>

kayabaNerve commented on coins/monero/src/wallet/send/builder.rs:
```diff
@@ -48,6 +50,13 @@ impl SignableTransactionBuilderInternal {
+  fn add_decoy(&mut self, decoy: Decoys) {
+    self.decoys.push(decoy);
+  }
+  fn add_decoys(&mut self, decoys: &[Decoys]) {
```
Why not add these with inputs?

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -147,27 +147,14 @@ pub enum TransactionError {
-  rpc: &Rpc<RPC>,
-  ring_len: usize,
+fn prepare_inputs(
   inputs: &[SpendableOutput],
```
Why not `(SpendableOutput, Decoys)`?

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -197,6 +184,55 @@ async fn prepare_inputs<R: RngCore + CryptoRng, RPC: RpcConnection>(
+  fee_rate: Fee,
+) -> (usize, u64) {
+  let mut weight: usize = 0;
+  let mut fee: u64 = 0;
```
FYI, 0usize, 0u64 are valid Rust.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -197,6 +184,55 @@ async fn prepare_inputs<R: RngCore + CryptoRng, RPC: RpcConnection>(
+
+    fee = fee_calculated_from_weight;
+
+    debug_assert!(iters < max_iters, "Reached max fee calculation attempts");
```
This isn't an actual invariant and shouldn't be a debug assert.

j-berman replied:
This is there to pick up an unexpected result. It should never happen unless there is a mistake in the code. Plus I figure an actual invariant should be an assert. What is `debug_assert` for if not for a situation like this?

kayabaNerve replied:
Agreed it should never happen. Statement as an invariant requires effective certainty this won't be hit, no matter the arguments/variables at play. We allow custom fee rate specification, and that introduces a massive surface scope.

Unlikely != impossible without error.

You can likely turn this into an invariant by stating that the fee will only go up, and will stop going up once the VarInt hits its maximum length in bytes. Additionally, any increases in the fee requires the VarInt having gone up in bytes, and accordingly, the maximum execution depth is 10 (preventing endless execution). But then the max_iters has to be the execution depth which it should already halt on, making this debug_assert pointless.

The reasons to remove this debug_assert, and unbound iterations which may allow ridiculous values, are:

1) So we can keep other debug_asserts
2) Because the fee already has to be inspected before signing continues

If the concern is programs not doing #2, we should emphasize the API around doing so like Timelock.

As for this specific debug_assert, you can rewrite as
```
#[cfg(test)]
debug_assert
```
So it's caught in testing yet will never effect actual programs, compiled with debug_asserts or not.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -235,9 +278,11 @@ pub struct SignableTransaction {
   protocol: Protocol,
   r_seed: Option<Zeroizing<[u8; 32]>>,
   inputs: Vec<SpendableOutput>,
+  decoys: Vec<Decoys>,
```
Would it be better to do (SpendableOutput, Decoys) in a single vec? This is internal so I don't actually care, just noting.

j-berman replied:
I like this suggestion. Done in the latest.

kayabaNerve commented on processor/src/coins/monero.rs:
```diff
@@ -536,7 +549,7 @@ impl Coin for Monero {
   #[cfg(test)]
   async fn get_fee(&self) -> Self::Fee {
-    self.rpc.get_fee().await.unwrap()
+    self.rpc.get_fee(0).await.unwrap()
```
Is 0 highest or lowest priority? Lowest, right?

j-berman replied:
0 is lowest. FWIW, upon working through the v17 fixes, I noticed the CLI, RPC, and GUI all end up defaulting this to 1 in the normal case thanks to a separate code path in those wallets that adjusts the priority down to 1 if there is no backlog in the pool. I defaulted to 1 in the latest and included a TODO to implement an `adjust_priority`

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -276,6 +321,49 @@ impl Change {
+  assert_eq!(
+    payments
+      .iter()
+      .filter(|payment| match *payment {
```
You can use `any` to return if any payments in the iter are change. Just FYI.

j-berman replied:
Think you're aware, but just clarifying, it's testing that the count is exactly as expected here

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -332,56 +422,89 @@ impl SignableTransaction {
 
-    Ok(SignableTransaction { protocol, r_seed, inputs, payments, data, fee })
+    // Sanity check the change again after modifying
+    sanity_check_change_payment(&payments, change_address.is_some());
```
What's the concern here?

j-berman replied:
That someone will accidentally break this invariant when modifying the code in the future. I could make the code cleaner to avoid any value in adding this assert here, but figured this is solid to ship and it's worth me moving on

kayabaNerve commented on coins/monero/src/wallet/send/multisig.rs:
```diff
@@ -153,22 +145,7 @@ impl SignableTransaction {
-    )
-    .await
-    .map_err(TransactionError::RpcError)?;
+    let decoys = self.decoys.clone();
```
You shouldn't clone decoys out of self when self is in the Machine. Just access them via `.signable`.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -332,56 +422,89 @@ impl SignableTransaction {
+    }
+
+    // Collect payments in a container that includes a change output if a change address is provided
+    let mut payments: Vec<InternalPayment> =
```
I don't believe this type specification is needed.

kayabaNerve commented on coins/monero/tests/runner.rs:
```diff
@@ -74,6 +75,18 @@ pub async fn get_miner_tx_output(rpc: &Rpc<HttpRpc>, view: &ViewPair) -> Spendab
 }
 
+/// Makes sure the weight and fee match the expected calculation
+pub fn check_weight_and_fee(tx: &Transaction, fee_rate: Fee) {
```
This should also be run on TXs sent via monero-wallet-rpc, which we should have in lib.

j-berman replied:
See 4905e6fd3c2eaf6a2264bab2ce00ea2b7e1b8790

If https://github.com/monero-project/monero/pull/8882 is merged then that commit could be reverted

kayabaNerve replied:
Add a TODO in codebase for when 8882 is merged?

kayabaNerve replied:
Add a TODO in codebase for when 8882 is merged?

kayabaNerve reviewed (commented):
utACK. Overall LGTM.

j-berman commented:
- In revisions, I noticed a missing parameter to the `get_fee_estimate` RPC that impacts the fee returned when tx volume fluctuates and dynamic fees trigger (the `grace_blocks` parameter). In the latest, the default RPC  request now matches wallet2 as of the latest. FWIW fairly positive I would have caught this when testing this PR with mainnet and am still planning to test with mainnet. Can look into implementing a test to trigger this scenario too.
- Ack all your comments I didn't explicitly respond to. I implemented those comments in the latest.
- v17 tests should be working now.
- Also calling out that there is an `adjust_priority` function that executes by default in the CLI, RPC, and GUI wallets that automatically adjusts the priority of a user-provided priority downward if there is no backlog in the transaction pool (and the user-provided priority would cause an unnecessarily expensive fee). I added a TODO for a future PR.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -622,7 +721,7 @@ impl SignableTransaction {
+    let mut fee = self.inputs.iter().map(|(input, _)| input.commitment().amount).sum::<u64>();
     let mut tx_outputs = Vec::with_capacity(outputs.len());
     let mut ecdh_info = Vec::with_capacity(outputs.len());
     for output in &outputs {
```
L633 in develop, 732 in your PR at time of comment is

`Some(output.view_tag).filter(|_| matches!(self.protocol, Protocol::v16))`

This should be consolidated with Protocol::view_tags.

j-berman replied:
There was another one later in the file too, updated both.

kayabaNerve commented on coins/monero/src/lib.rs:
```diff
@@ -67,15 +67,36 @@ impl Protocol {
+
+  /// Whether or not the specified version uses the fee algorithm released in
+  /// Monero v18.
+  pub fn v18_fee(&self) -> bool {
```
In the future, it may make sense for the fee algorithm to be an enum. Since we don't plan to support legacy fee algorithms though, I'm not asking for it here.

kayabaNerve commented on coins/monero/src/lib.rs:
```diff
@@ -67,15 +67,36 @@ impl Protocol {
+  pub fn v18_fee(&self) -> bool {
+    match self {
+      Protocol::v14 => false,
+      Protocol::v16 => true,
```
Why do you have Protocol::v16 using the v18 fee algorithm?

j-berman replied:
I was referring to Monero release version v18, but agreed this was confusing. Renamed the fn to `v16_fee` for consistency.

kayabaNerve commented on coins/monero/src/ringct/bulletproofs/mod.rs:
```diff
@@ -35,23 +35,38 @@ pub enum Bulletproofs {
-        let size = (fields + (2 * LR_len)) * 32;
-        ((base * padded_outputs) - size) * 4 / 5
-      }
+    let mut bp_clawback = 0;
```
Cite the BP clawback algorithm with a link to Monero's?

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -18,6 +18,9 @@ use crate::{
 };
 
+// Number of blocks the fee estimate will be valid for
+const GRACE_BLOCKS_FOR_FEE_ESTIMATE: u64 = 10;
```
This should be the same constant as wallet2. Cite source?

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +520,69 @@ impl<R: RpcConnection> Rpc<R> {
   /// MUST be sanity checked.
   // TODO: Take a sanity check argument
-  pub async fn get_fee(&self) -> Result<Fee, RpcError> {
+  pub async fn get_fee(&self, protocol: Protocol, priority: u32) -> Result<Fee, RpcError> {
```
Should priority be a u32 or an enum? Match Protocol and do enum with Custom(u32)?

Also, just double checking it's u32 and not i32.

j-berman replied:
Implemented an enum.

It's u32: https://github.com/monero-project/monero/blob/94e67bf96bbc010241f29ada6abc89f49a81759c/src/wallet/wallet2.cpp#L7610

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +520,69 @@ impl<R: RpcConnection> Rpc<R> {
+        .await?;
+
+      let priority_idx =
+        if priority >= 4 { 3 } else { priority.saturating_sub(1).try_into().unwrap() };
```
Error InvalidPriority?

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +520,69 @@ impl<R: RpcConnection> Rpc<R> {
+
+      if res.status != "OK" {
+        Err(RpcError::InvalidFee)
+      } else if priority_idx >= res.fees.len() {
```
`res.fees.get(priority.try_into().unwrap())`, moving the `.try_into().unwrap()` off saturating sub for clarity, then `.ok_or(RpcError::InvalidPriority)?` for Option to Result.

j-berman replied:
Leaving this one for you

kayabaNerve commented on coins/monero/src/serialize.rs:
```diff
@@ -7,8 +7,18 @@ use curve25519_dalek::{
-pub(crate) fn varint_len(varint: usize) -> usize {
-  ((usize::try_from(usize::BITS - varint.leading_zeros()).unwrap().saturating_sub(1)) / 7) + 1
+mod sealed {
+  pub trait Varint: TryInto<u64> {}
```
VarInt

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +520,69 @@ impl<R: RpcConnection> Rpc<R> {
 
-    let res: FeeResponse = self.json_rpc_call("get_fee_estimate", None).await?;
-    Ok(Fee { per_weight: res.fee, mask: res.quantization_mask })
+    let priority_idx = if priority == 0 { 1 } else { priority - 1 };
```
So 0 -> 1, and 2 -> 1, yet 1 -> 0? Please document this monstrosity better, ideally citing Monero RPC docs yet alternatively the code it uses for this.

j-berman replied:
Cited the code.

kayabaNerve commented on coins/monero/src/rpc.rs:
```diff
@@ -513,16 +520,69 @@ impl<R: RpcConnection> Rpc<R> {
+      // though not an RPC error, it seems sensible to treat as such
+      Err(RpcError::InvalidPriority)?;
+    }
+    let fee_multiplier = multipliers[priority_idx as usize];
```
try_into, please.

kayabaNerve commented on coins/monero/src/transaction.rs:
```diff
@@ -331,4 +339,26 @@ impl Transaction {
   }
+
+  fn is_rct_bulletproof(&self) -> bool {
+    matches!(self.rct_signatures.prunable.rct_type(), 3 | 4 | 5)
```
Can we match prunable, returning true/false for each possible enum variant?

j-berman replied:
Used the nice new `rct_signatures.rct_type()` in the latest.

kayabaNerve commented on coins/monero/src/transaction.rs:
```diff
@@ -331,4 +339,26 @@ impl Transaction {
+  }
+
+  fn is_rct_bulletproof_plus(&self) -> bool {
+    matches!(self.rct_signatures.prunable.rct_type(), 6)
```
Same comment as is_rct_bulletproof.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -33,7 +33,7 @@ use crate::{
   },
   transaction::{Input, Output, Timelock, TransactionPrefix, Transaction},
-  rpc::{RpcError, RpcConnection, Rpc},
+  rpc::{RpcError},
```
Unnecessary curly bois

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -276,6 +318,49 @@ impl Change {
+}
+
+fn sanity_check_change_payment(payments: &[InternalPayment], has_change_address: bool) {
+  assert_eq!(
```
debug_assert unless you're concerned over an API invariant.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -332,62 +418,98 @@ impl SignableTransaction {
-    if in_amount < out_amount {
-      Err(TransactionError::NotEnoughFunds(in_amount, out_amount))?;
+    let in_amount = inputs.iter().map(|(input, _)| input.commitment().amount).sum::<u64>();
+    if in_amount < out_amount + fee {
```
Parens on rhs, please.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -332,62 +418,98 @@ impl SignableTransaction {
+    sanity_check_change_payment(&payments, change_address.is_some());
+
+    // Sanity check outgoing amount + fee == incoming amount
+    assert_eq!(
```
debug_assert?

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -419,31 +541,8 @@ impl SignableTransaction {
-    }
-    let modified_change_ecdh = subaddresses && (!additional);
+    let (subaddresses, additional) = need_additional(payments);
+    let modified_change_ecdh = subaddresses && !additional;
```
Parens on right term, please.

kayabaNerve commented on coins/monero/src/wallet/send/mod.rs:
```diff
@@ -286,10 +371,11 @@ impl SignableTransaction {
     protocol: Protocol,
     r_seed: Option<Zeroizing<[u8; 32]>>,
-    inputs: Vec<SpendableOutput>,
+    inputs: Vec<(SpendableOutput, Decoys)>,
```
These decoys need to be checked to have the proper ring_len.

kayabaNerve commented on coins/monero/src/wallet/send/multisig.rs:
```diff
@@ -148,36 +134,17 @@ impl SignableTransaction {
       clsags.push(AlgorithmMachine::new(clsag, offset));
     }
 
-    // Select decoys
```
Transcript the decoys since they're no longer deterministic to the transcript, please.

kayabaNerve commented on coins/monero/tests/send.rs:
```diff
@@ -116,3 +148,126 @@ test!(
+    |_, tx: Transaction, _, mut state: (Scanner, Vec<SubaddressIndex>)| async move {
+      let mut scanned_tx = state.0.scan_transaction(&tx).not_locked();
+
+      use std::collections::HashMap;
```
Import at the top?

j-berman replied:
I moved to the top of that function in the latest (since top of the file gives unused warnings). How's that? Or is the preference to suppress the warning?

kayabaNerve replied:
Top of block is good.

kayabaNerve commented on processor/src/coins/monero.rs:
```diff
@@ -398,6 +399,28 @@ impl Coin for Monero {
+    .unwrap();
+
+    let inputs: Vec<(SpendableOutput, Decoys)> =
+      spendable_outputs.into_iter().zip(decoys.into_iter()).collect();
```
`collect::<Vec<_>>();`

kayabaNerve commented on processor/src/coins/monero.rs:
```diff
@@ -583,17 +598,32 @@ impl Coin for Monero {
+    .unwrap();
+
+    let inputs: Vec<(SpendableOutput, Decoys)> =
+      outputs.into_iter().zip(decoys.into_iter()).collect();
```
`collect::<Vec<_>>();`

kayabaNerve reviewed (commented):
Very thankful to have this :D

kayabaNerve commented:
CI failure due to #291.

kayabaNerve committed:
Minor lint
